src/jose/jws/general/verifier.ts

```diff
@@ -0,0 +1,112 @@
+import { DIDResolver } from '../../../did/did-resolver';
+import { verify as verifyEd25519 } from '../../algorithms/ed25519';
+import { verify as verifySecp256k1 } from '../../algorithms/secp256k1';
+
+const verifiers: { [key:string]: VerifyFn } = {
+  'Ed25519'   : verifyEd25519,
+  'secp256k1' : verifySecp256k1
+};
+
+type VerificationResult = {
+  signers: string[];
+};
+
+// TODO: add logic to prevent validating duplicate signatures
+export class GeneralJwsVerifier {
+  jws: GeneralJws;
+
+  constructor(jws: GeneralJws) {
+    this.jws = jws;
+  }
+
+  async verify(didResolver: DIDResolver): Promise<VerificationResult> {
+    const signers: string[] = [];
+
+    for (let signature of this.jws.signatures) {
+      const protectedBytes = base64url.baseDecode(signature.protected);
+      const protectedJson = new TextDecoder().decode(protectedBytes);
+
+      const { kid } = JSON.parse(protectedJson);
+      const did = this.extractDid(kid);
+      const publicKeyJwk = await this.getPublicKey(did, kid, didResolver);
+
+      const isVerified = await this.verifySignature(this.jws.payload, signature, publicKeyJwk);
+
+      if (isVerified) {
+        signers.push(did);
+      } else {
+        throw new Error('signature verification failed');
+      }
+    }
+
+    return { signers };
+  }
+
+  async getPublicKey(did: string, kid: string, didResolver: DIDResolver): Promise<JwkPublic> {
+    // `resolve` throws exception if DID is invalid, DID method is not supported,
+    // or resolving DID fails
+
+    const { didDocument } = await didResolver.resolve(did);
+    const { verificationMethod: verificationMethods = [] } = didDocument || {};
+
+    let verificationMethod: VerificationMethod | undefined;
+
+    for (const vm of verificationMethods) {
+      // consider optimizing using a set for O(1) lookups if needed
+      if (vm.id === kid) {
+        verificationMethod = vm;
+        break;
+      }
+    }
+
+
+    if (!verificationMethod) {
+      throw new Error('public key needed to verify signature not found in DID Document');
+    }
+
+    // TODO: replace with JSON Schema based validation
+    // more info about the `JsonWebKey2020` type can be found here:
+    // https://www.w3.org/TR/did-spec-registries/#jsonwebkey2020
+    if (verificationMethod.type !== 'JsonWebKey2020') {
+      throw new Error(`verification method [${kid}] must be JsonWebKey2020`);
+    }
+
+    const { publicKeyJwk } = verificationMethod;
+
+    // TODO: replace with JSON Schema based validation
+    // more info about the `publicKeyJwk` property can be found here:
+    // https://www.w3.org/TR/did-spec-registries/#publickeyjwk
+    if (!publicKeyJwk) {
+      throw new Error(`publicKeyJwk property not found on verification method [${kid}]`);
+    }
+
+    return publicKeyJwk as JwkPublic;
+  }
+
+  async verifySignature(base64UrlPayload: string, signature: Signature, jwkPublic: JwkPublic): Promise<boolean> {
+    const verifyFn: VerifyFn = verifiers[jwkPublic.crv];
```

Comment:
Prefer naming style consistency throughout code base.
```suggestion
    const verifyFunction: VerifyFunction = verifiers[jwkPublic.crv];
```

Reply:
Thoughts on `Verifier` instead? I actually have formed the opinion that signing should imply verification. This should be combined into an interface type called `Signer` that algorithms implement. Thoughts?

Reply:
Make sense, I thought the reason you didn't want to have algorithms implement an interface so that we don't need to instantiate instances to use those functions?

Reply:
went a slightly different route but it did:
* address the abbreviation
* implement an interface without the need to instantiate
* reduce the number of files that need to be touched to add a new algo down to 2 (though i'm not sure how often we'll be adding keys)
* prepared the jose dir for encryption-related algos


you may like it less but i think it should work for now 🤣 https://github.com/TBD54566975/dwn-sdk-js/commit/2456261f3856e561e5af7b67190dd02a598faa33

Reply:
This will do! This code has under gone 900 iteration of abstractions. = ]

By the way, REALLY love how you address each comment as individual commits, makes reviewer (my) job so much easier!